tests/test_fetch.py

```diff
@@ -347,3 +348,33 @@ def __init__(self, status_code):
         "https://example.org/file"
     ):
         raise HTTPError(response=Response(401))
+
+
+@pytest.mark.parametrize("sha256_type", ("original", "upper", "gibberish", "bad-type"))
+def test_checksum_checks_bytes(
+    tmp_path: Path, package_repository_base, package_server, sha256_type
+):
+    host, port = package_server.getsockname()
+    base = f"http://{host}:{port}/test"
+    package_name = "zlib-1.2.11-h7b6447c_3.conda"
+    url = f"{base}/linux-64/{package_name}"
+    package_path = package_repository_base / "linux-64" / package_name
+    sha256 = checksum(package_path, algorithm="sha256")
+    size = package_path.stat().st_size
+    output_path = tmp_path / package_name
+
+    if sha256_type in ("gibberish", "bad-type"):
+        with pytest.raises(CondaValueError):
+            download(
+                url,
+                output_path,
+                size=size,
+                sha256=123456 if sha256_type == "bad-type" else "not-an-hex-string",
+            )
+    else:
+        download(
+            url,
+            output_path,
+            size=size,
+            sha256=sha256.upper() if sha256_type == "upper" else sha256,
+        )
```

Comment:
I prefer using a `contextlib.nullcontext` instead of repeating the download logic:

```
@pytest.mark.parametrize(
    "raises,get_sha256",
    [
        pytest.param(False, lambda x: x, id="original"),
        pytest.param(False, str.upper, id="upper"),
        pytest.param(True, lambda x: "not-an-hex-string", id="gibberish"),
        pytest.param(True, lambda x: 123456, id="bad-type"),
    ],
)
def test_checksum_checks_bytes(
    tmp_path: Path,
    package_repository_base,
    package_server,
    raises: bool,
    get_sha256: Callable[[str], Any],
):
    host, port = package_server.getsockname()
    base = f"http://{host}:{port}/test"
    package_name = "zlib-1.2.11-h7b6447c_3.conda"
    url = f"{base}/linux-64/{package_name}"
    package_path = package_repository_base / "linux-64" / package_name
    sha256 = checksum(package_path, algorithm="sha256")
    size = package_path.stat().st_size
    output_path = tmp_path / package_name

    with pytest.raises(CondaValueError) if raises nullcontext():
        download(url, output_path, size=size, sha256=get_sha256(sha256))
```

Reply:
Oh, nice.  Thanks! Committed the result.